File: 1753.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

import heapq

INF = int(1e9)

def dijkstra(distance, graph, K):
  queue = []
  heapq.heappush(queue, (0,K))
  distance[K] = 0

  while queue:
    dist, current = heapq.heappop(queue)
    if distance[current] < dist:
      continue
    for adj in graph[current]:
      if distance[adj[0]] > adj[1] + distance[current]:
        distance[adj[0]] = adj[1] + distance[current]
        heapq.heappush(queue, (distance[adj[0]], adj[0]))
  return distance

V, E = map(int, input().split())
K = int(input())

distance = [INF for i in range(V+1)]
graph = [[] for i in range(V+1)]

for _ in range(E):
    start, end, dist = map(int, input().split())
    graph[start].append((end,dist))

distance = dijkstra(distance, graph, K)
for i in range(1, V+1):
  if distance[i] == INF:
    print("INF")
  else:
    print(distance[i])

File: 1753.java
import java.io.*;
import java.util.*;

public class Main {
    
    static class Node{
        int num;
        int weight;
        
        public Node(int num, int weight) {
            this.num = num;
            this.weight = weight;
        }
    }
    
    public static int V,E,K;
    public static ArrayList<ArrayList<Node>> graph;
    public static int[] dist;
    
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        String[] info = br.readLine().split(" ");
        V = Integer.parseInt(info[0]);
        E = Integer.parseInt(info[1]);
        K = Integer.parseInt(br.readLine());
        
        graph = new ArrayList<ArrayList<Node>>();
        for(int i=0; i<=V; i++) {
            graph.add(new ArrayList<Node>());
        }
        
        for(int i=0; i<E; i++) {
            String[] connection = br.readLine().split(" ");
            int u = Integer.parseInt(connection[0]);
            int v = Integer.parseInt(connection[1]);
            int w = Integer.parseInt(connection[2]);
            
            graph.get(u).add(new Node(v,w));
        }
        
        dist = new int[V+1];
        for(int i=0; i<=V; i++) {
            dist[i] = Integer.MAX_VALUE;
        }
        
        PriorityQueue<Node> queue = new PriorityQueue<Node>((o1,o2) -> Integer.compare(o1.weight, o2.weight));
        queue.offer(new Node(K,0));
        dist[K] = 0;
        
        while(!queue.isEmpty()) {
            Node current = queue.poll();
            
            if(dist[current.num] < current.weight) {
                continue;
            }
            
            for(int i=0; i<graph.get(current.num).size(); i++) {
                Node next = graph.get(current.num).get(i);
                
                if(dist[next.num] > current.weight + next.weight) {
                    dist[next.num] = current.weight + next.weight;
                    queue.offer(new Node(next.num, dist[next.num]));
                }
            }
        }
        
        for(int i=1; i<=V; i++) {
            if(dist[i] == Integer.MAX_VALUE) {
                bw.write("INF\n");
            }
            else {
                bw.write(String.valueOf(dist[i]) + "\n");
            }
        }
        bw.flush();
        bw.close();
    }
}
